### PR TITLE
feat: support experimental loader `emitDiagnostic`

### PR DESCRIPTION
**Pull request by h-a-n-a**
## Summary

closes https://github.com/web-infra-dev/rspack/issues/5913

Supported experimental loader API `emitDiagnostic`. 

This works almost the same as `emitError` or `emitWarning`. But, it's for emitting rich-formatted diagnostics.

```ts
export interface DiagnosticLocation {
	/** Text for highlighting the location */
	text?: string;
	/** 1-based line */
	line: number;
	/** 0-based column in bytes */
	column: number;
	/** Length in bytes */
	length: number;
}

export interface Diagnostic {
	message: string;
	help?: string;
	sourceCode?: string;
	/**
	 * Location to the source code.
	 *
	 * If `sourceCode` is not provided, location will be omitted.
	 */
	location?: DiagnosticLocation;
	file?: string;
	severity: "error" | "warning";
}

interface LoaderExperiments {
	emitDiagnostic(diagnostic: Diagnostic): void;
}
```

## Example

For loader:

```js
/** @type {import("@rspack/core").LoaderDefinition} */
module.exports = function() {
	this.experiments.emitDiagnostic({
		message: "`React` is not defined",
		severity: "error",
		sourceCode: `<div></div>`,
		help: "try to import `React`",
		location: {
			line: 1,
			column: 1,
			length: 3,
		}
	});
	return ""
}
```

it outputs:

```
ERROR in (./with-help.js!) 1:1-4
  × ModuleError: `React` is not defined
   ╭────
 1 │ <div></div>
   ·  ───
   ╰────
  help: try to import `React`
```

Note: This is an **experimental** API and is rspack-specific.

## Checklist

- [X] Tests updated (or not required).
- [ ] Documentation updated (or not required).

### CODE REVIEW

File: crates/node_binding/src/diagnostic.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        location.column as usize,
+        location.length as usize,
+      )
+      .ok_or_else(|| {
+        Error::new(
+          Status::Unknown,
+          "Format diagnostic failed: Invalid location. Did you pass the correct line, column and length?",
+        )
+      })?;
+      let end_byte = offset.saturating_add(length);
```
Comment (hardfist) [inline crates/node_binding/src/diagnostic.rs:74]: > or >= ?

File: packages/rspack-test-tools/tests/diagnosticsCases/module-build-failed/loader-emit-diagnostic/stats.err [hunk 2]
```diff
@@ -0,0 +1,102 @@
+WARNING in (./basic.js!)
+  ⚠ ModuleWarning: `React` is not defined
+
```
Comment (hardfist) [inline packages/rspack-test-tools/tests/diagnosticsCases/module-build-failed/loader-emit-diagnostic/stats.err:3]: maybe the file is called basic-loader.js is more readable here 🤣

File: crates/node_binding/src/diagnostic.rs (hunk 1 above)
Comment (h-a-n-a) [reply crates/node_binding/src/diagnostic.rs:74]: Did you mean `end_byte >= rope.len_bytes()`?

The largest byte index can be the length of the source code in bytes (valid if end_byte <= rope.len_bytes()). For example, for code snippet "a", the offset could be 0 and length could be 1. This is still a valid label span.

File: crates/node_binding/test.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+const { formatDiagnostic } = require("./binding")
```
Comment (hardfist) [inline crates/node_binding/test.js:1]: this file is used only for local testing? or should we run real unit test for it

Comment (h-a-n-a) [reply crates/node_binding/test.js:1]: This is for local testing only. Let me remove this.

**Comment by hardfist** [comment]
waiting for @colinaaa's feedback on this

**Comment by colinaaa** [comment]
It looks great to me! Thank you @h-a-n-a @hardfist

**Review by hardfist — APPROVED** [review]

**Pull request merged**